namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala

```diff
@@ -32,80 +30,84 @@ class DnsSrvNamer(
+        val vaddr = watchDns(address).run.map {
+          case Activity.Ok(rsp) =>
+            Addr.Bound(rsp: _*)
+          case Activity.Pending =>
+            Addr.Pending
+          case Activity.Failed(e) =>
+            log.debug("SDS lookup failure: %s", e.getMessage)
+            Addr.Failed(e)
+        }
 
-          val done = new AtomicBoolean(false)
-          val initialized = new AtomicBoolean(false)
+        val state: Var[Activity.State[NameTree[Name]]] = vaddr.map {
+          case Addr.Bound(addrs, _) if addrs.isEmpty =>
+            Activity.Ok(NameTree.Neg)
+          case Addr.Bound(addrs, _) =>
+            Activity.Ok(NameTree.Leaf(Name.Bound(vaddr, prefix ++ id, path.drop(1))))
+          case Addr.Pending =>
+            Activity.Pending
+          case Addr.Failed(_) =>
+            Activity.Ok(NameTree.Neg)
+        }
 
-          Future.whileDo(!done.get) {
-            lookupSrv(address, prefix ++ id, path.drop(1)).transform { result =>
-              result match {
-                case Return(nameTree) =>
-                  initialized.set(true)
-                  state.update(Activity.Ok(nameTree))
-                case Throw(e) =>
-                  failure.incr()
-                  log.error(e, "resolution error: %s", address)
-                  if (!initialized.get) {
-                    state.update(Activity.Failed(e))
-                  }
-              }
-              Future.sleep(refreshInterval)
-            }
-          }
-
-          Closable.make { _ =>
-            done.set(true)
-            Future.Unit
-          }
-
-        })
-      case _ => Activity.value(NameTree.Neg)
+        Activity(state)
+      case _ =>
+        Activity.value(NameTree.Neg)
     }
   }
 
-  private def lookupSrv(address: String, id: Path, residual: Path): Future[NameTree[Name]] = {
-    log.debug("looking up %s", address)
-    pool {
-      val lookup = new DNS.Lookup(address, DNS.Type.SRV, DNS.DClass.IN)
-      lookup.setResolver(resolver)
-      Stat.time(latency)(lookup.run())
-      lookup.getResult match {
-        case DNS.Lookup.HOST_NOT_FOUND | DNS.Lookup.TYPE_NOT_FOUND =>
-          log.trace("no results for %s", address)
-          failure.incr()
-          NameTree.Neg
-        case DNS.Lookup.SUCCESSFUL =>
-          val answers = Option(lookup.getAnswers).getOrElse(Array.empty)
-          val srvRecords = answers.flatMap {
-            case srv: DNS.SRVRecord => try {
-              val inetAddress = InetAddress.getByName(srv.getTarget.toString())
-              Some(Address(new InetSocketAddress(inetAddress, srv.getPort)))
-            } catch {
-              case _: UnknownHostException =>
-                log.warning(s"srv lookup of $address returned unknown host ${srv.getTarget}")
-                badHosts.incr()
-                None
+  private def watchDns(dsnsrvRecord: String)(implicit timer: Timer = DefaultTimer): Activity[List[Address]] = {
```

Comment:
I don't think you need to provide `DefaultTimer` here; there should be an implicit timer in scope for the whole class -- see line 20